deepspeed/utils/numa.py

```diff
@@ -121,11 +121,65 @@ def get_numactl_cmd(bind_core_list, num_local_procs, local_rank):
     # check if all cores belong to same numa, if true, bind process to that numa domain with -m parameter
     numa_cores = get_numa_cores()
     num_numas = len(numa_cores)
+
+    numa_mode = "normal"
+
+    non_empty_numa_list = []
+    empty_numa_list = []
+    previous_numa_cores = []
+    numa_node_list = []
+    numa_node_list_list = []
     for i in range(num_numas):
-        if set(core_list_for_rank) <= set(numa_cores[i]):
-            numactl_cmd.append("-m")
-            numactl_cmd.append(f"{i}")
-            break
+        # look for empty numa which is HBM numa
+        if numa_cores[i] == []:
+            empty_numa_list.append(i)
+        else:
+            non_empty_numa_list.append(i)
+
+            # check for fakenuma
+            if numa_cores[i] == previous_numa_cores:
+                if numa_node_list == []:
+                    #first duplication, add previous node into list
+                    numa_node_list.append(i - 1)
+                numa_node_list.append(i)
+            else:
+                if numa_node_list != []:
+                    numa_node_list_list.append(numa_node_list)
+                    numa_node_list = []
+        previous_numa_cores = numa_cores[i]
+    if numa_node_list != []:
+        numa_node_list_list.append(numa_node_list)
+
+    if empty_numa_list != [] and len(empty_numa_list) == len(non_empty_numa_list):
+        numa_mode = "flat_hbm"
+        numa_dict = dict(zip(non_empty_numa_list, empty_numa_list))
+    elif numa_node_list_list != []:
+        numa_mode = "fake"
+
+    if numa_mode == "normal":
+        for i in range(num_numas):
+            if set(core_list_for_rank) <= set(numa_cores[i]):
+                numactl_cmd.append("-m")
+                numactl_cmd.append(f"{i}")
+                break
+    elif numa_mode == "flat_hbm":
+        for i in range(num_numas):
+            if set(core_list_for_rank) <= set(numa_cores[i]):
+                numactl_cmd.append("-p")
```

Comment:
@tjruwase I have a question.  Sometimes I get request from engineers that in the case of HBM flatmode, they want to manually specific NUMA preference.  Mostly one of the following two scenarios:
1. Use -m instead of -p to make sure the workload would stay in HBM.  The workload may get an OOM when HBM memory is used up, but high bandwidth is ensured and workload won't allocate from DDR memory silently.
2. Prefer DDR NUMA node instead of HBM NUMA node, because DDR has lower latency than HBM, for smaller workload DDR might be better than HBM memory.

I think the default behavior should make all memory available, this is why `-p` HBM node is used for HBM flat mode.   We want to provide different choice for different user with different needs.   We have two options:

1. Use an environment variable for user to choose numa mode under HBM flat mode.
2. Use a deepspeed command line option to choose.

Which one is better, do you have any suggestions?  Thanks!

Reply:
Update: we tested small models and found that even when model as small as 6B parameters, HBM still has better performance than DDR.  So -p should have good generality.  Will mark this as resolved.